### ISSUE TITLE
withAPIData does not work with custom endpoints

### ISSUE THREAD

**Issue by donnapep** [WordPress/gutenberg#3209]
## Issue Overview
`withAPIData` does not work with custom endpoints.

## Steps to Reproduce (for bugs)

1. Install and activate [Sensei](https://github.com/automattic/sensei/tree/update/block-support) (you’ll need the `update/block-support`​ branch)
2. Download [Sensei Courses](https://github.com/automattic/sensei-courses/tree/update/use-custom-endpoint) (you'll need the `update/use-custom-endpoint` branch). `cd` into the directory and run `npm install` and `npm run build`. Upload the plugin to your site and activate it.
3. Create a course by going to _Courses_ > _Add New_. Give your course a title and description.
4. Create a new page.
5. Add the _Courses_ block.
6. Watch the _Network_ tab in Chrome DevTools. Notice that there is no request being sent to `/sensei/v1/courses`, although there is a request to `/wp/v2/course-category`.

Chrome 61.0.3163.100 on MacOS 10.12.6 and WordPress 4.9-beta4-42028, Gutenberg 1.5.2.

## Expected Behavior

I expect a request to be made to the custom endpoint and results to be returned.

## Current Behavior

No API data is returned. In fact, the request is not even sent.

## Possible Solution

When using a `path` set to `/sensei/v1/courses`, the [`getRoute` function](https://github.com/WordPress/gutenberg/blob/master/components/higher-order/with-api-data/routes.js#L50) returns `false`, which in turn causes the `applyMapping` function to [return early](https://github.com/WordPress/gutenberg/blob/master/components/higher-order/with-api-data/index.js#L173).

In other words, it seems like there is currently a requirement for a custom endpoint to start with `/wp/v2`.

## Todos
- [ ] Tests
- [ ] Documentation

**Comment by ockham** [comment]
> When using a `path` set to `/sensei/v1/courses`, the [`getRoute` function](https://github.com/WordPress/gutenberg/blob/master/components/higher-order/with-api-data/routes.js#L50) returns `false`, which in turn causes the `applyMapping` function to [return early](https://github.com/WordPress/gutenberg/blob/master/components/higher-order/with-api-data/index.js#L173).

> In other words, it seems like there is currently a requirement for a custom endpoint to start with `/wp/v2`.

This seems to come from the `window.wpApiSettings` global (thru https://github.com/WordPress/gutenberg/blob/1cbe5e6a719e2b3c701e56a39452769762ddc935/components/higher-order/with-api-data/index.js#L26 and https://github.com/WordPress/gutenberg/blob/b53887488b2de0ff28a9aa9c6e05f5b4c3c95af2/editor/provider/index.js#L118 among others), which includes a `schema` attr that looks like this for me:

![image](https://user-images.githubusercontent.com/96308/32614363-62f861d6-c56d-11e7-87d8-f55509e99be5.png)

This really doesn't include any `/sensei` routes.

/cc @aduth

**Comment by youknowriad** [comment]
I guess that's probably the reason we need the `init` call for our custom namespace here https://github.com/WordPress/gutenberg/pull/3377/files#diff-620130744b6b73b822a3df609671d930R776

So that might be the way to fix this for custom endpoints.

**Comment by aduth** [comment]
Hmm, ideally we'd not have to initialize each namespace separately and could instead load in a single schema for all endpoints of the REST API.

Might be that we drop `/wp/v2` to `/` here? Untested theory:

https://github.com/WordPress/gutenberg/blob/05f8169f7ff3880c70b48cb3a1dd1c58ec1b33b6/lib/client-assets.php#L480-L487

**Comment by greatislander** [comment]
Related: #3462
